### PR TITLE
Update to new command api

### PR DESCRIPTION
**Pull request by grant0417**

### CODE REVIEW

File: src/copilot.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import YAML from "yaml";
 
 const applicationName: Fig.Generator = {
-  script: "cat copilot/.workspace",
+  script: ["cat","copilot/.workspace"],
```
Comment (chaynabors) [inline src/copilot.ts:4]: nit: spacing

File: src/cordova.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ const commonOptions: Fig.Option[] = [
 ];
 
 const platformGenerator: Fig.Generator = {
-  script: "cat package.json",
+  script: ["cat","package.json"],
```
Comment (chaynabors) [inline src/cordova.ts:27]: nit: spacing

File: src/docker.ts [hunk 3]
```diff
@@ -346,10 +353,10 @@ const sharedCommands: Record<string, Fig.Subcommand> = {
                 fileFlagIndex = context.indexOf("--file");
                 dockerfilePath = context[fileFlagIndex + 1];
               } else {
-                dockerfilePath = "$PWD/Dockerfile";
+                dockerfilePath = "Dockerfile";
               }
 
-              return `\\grep -iE 'FROM.*AS' "${dockerfilePath}"`;
+              return ["grep", "-iE", "FROM.*AS", dockerfilePath];
```
Comment (chaynabors) [inline src/docker.ts:359]: Is this right?

File: src/doppler.ts [hunk 4]
```diff
@@ -64,7 +64,7 @@ const projectsGenerator: Fig.Generator = {
     cacheKey: "projects",
     cacheByDirectory: true,
   },
-  script: "doppler projects --json",
+  script: ["doppler","projects","--json"],
```
Comment (chaynabors) [inline src/doppler.ts:67]: nit: spacing

File: src/dotnet/dotnet-new.ts [hunk 5]
```diff
@@ -7,7 +7,13 @@ type SearchResultData = {
 const searchGenerator: Fig.Generator = {
   script(context) {
     const searchTerm = context[context.length - 1];
-    return `curl -s -H "Accept: application/json" "https://azuresearch-usnc.nuget.org/query?packageType=Template&q=${searchTerm}"`;
+    return [
+      "curl",
+      "-sfL",
```
Comment (chaynabors) [inline src/dotnet/dotnet-new.ts:12]: This is different

File: src/dotnet/dotnet-run.ts [hunk 6]
```diff
@@ -58,7 +58,7 @@ const completionSpec: Fig.Spec = {
         name: "name",
         suggestions: ["Development", "Staging", "Production"],
         generators: {
-          script: "cat Properties/launchSettings.json",
+          script: ["cat","Properties/launchSettings.json"],
```
Comment (chaynabors) [inline src/dotnet/dotnet-run.ts:61]: nit: spacing

File: src/elm-json.ts [hunk 7]
```diff
@@ -1,8 +1,13 @@
 import { filepaths } from "@fig/autocomplete-generators";
 
 const packageList: Fig.Generator = {
-  script:
-    "curl -sH 'accept-encoding: gzip' --compressed https://package.elm-lang.org/search.json",
+  script: [
+    "curl",
+    "-sH",
+    "accept-encoding: gzip",
```
Comment (chaynabors) [inline src/elm-json.ts:7]: Not sure if this gets evaluated the same as before

File: src/ffmpeg.ts [hunk 8]
```diff
@@ -413,7 +413,7 @@ const completionSpec: Fig.Spec = {
       args: {
         name: "codec",
         generators: {
-          script: "ffmpeg -codecs",
+          script: ["ffmpeg","-codecs"],
```
Comment (chaynabors) [inline src/ffmpeg.ts:416]: nit: spacing

File: src/fig/shared.ts [hunk 9]
```diff
@@ -451,7 +483,7 @@ const scriptOptions = (script: ScriptFields) => {
           generators = param?.selector?.generators
             .filter((generator) => generator.type === "ShellScript")
             .map((generator) => ({
-              script: generator?.shellScript?.script,
+              script: ["bash", "-c", generator?.shellScript?.script],
```
Comment (chaynabors) [inline src/fig/shared.ts:486]: this looks shady

File: src/firebase.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
 const projectAliasesGenerator: Fig.Generator = {
-  script: "firebase projects:list", // this calls to a firebase server and is therefore slow
+  script: ["firebase","projects:list"], // this calls to a firebase server and is therefore slow
```
Comment (chaynabors) [inline src/firebase.ts:2]: nit: spacing

File: src/fisher.ts [hunk 11]
```diff
@@ -70,7 +70,7 @@ const pluginList = [
 ];
 
 const installedPlugins: Fig.Generator = {
-  script: "fish -c 'fisher list'",
+  script: ["fish","-c","fisher list"],
```
Comment (chaynabors) [inline src/fisher.ts:73]: nit: spacing

File: src/gem.ts [hunk 12]
```diff
@@ -2,13 +2,19 @@ const gems: Fig.Generator = {
...
+    const { stdout } = await executeShellCommand({
+      command: "gem",
+      args: [
+        "search",
+        "--both",
+        "--no-versions",
+        "--no-details",
+        "--quiet",
+        "--norc",
+        searchTerm,
```
Comment (chaynabors) [inline src/gem.ts:14]: this no longer replaces

File: src/gh.ts [hunk 13]
```diff
@@ -89,15 +89,24 @@ const ghGenerators: Record<string, Fig.Generator> = {
...
-        `gh repo list ${userOrOrg} --limit 9999 --json "nameWithOwner,description,isPrivate" `
-      );
+      const { stdout, status } = await execute({
+        command: "gh",
+        args: [
+          "repo",
+          "list",
+          userOrOrg,
+          "--limit",
+          "9999",
```
Comment (chaynabors) [inline src/gh.ts:99]: should probably cap this

File: src/hugo.ts [hunk 14]
```diff
@@ -874,7 +874,7 @@ const completionSpec: Fig.Spec = {
             {
               name: "archetype|default",
               generators: {
-                script: "ls ./archetypes/",
+                script: ["ls","./archetypes/"],
```
Comment (chaynabors) [inline src/hugo.ts:877]: nit: spacing

File: src/ignite-cli.ts [hunk 15]
```diff
@@ -1,5 +1,5 @@
 const generatorsGenerator: Fig.Generator = {
-  script: "ls ignite/templates",
+  script: ["ls","ignite/templates"],
```
Comment (chaynabors) [inline src/ignite-cli.ts:2]: nit: spacing

File: src/lerna.ts [hunk 16]
```diff
@@ -27,10 +25,14 @@ const getBranches: Fig.Generator = {
 
 const getAllScriptsFromPackages: Fig.Generator = {
   // Get all lerna packages, loop over them and get content of package.json
-  script: `lerna list -p | while read p; do\n \\cat $p/package.json && echo ${SPLIT_CHAR}\ndone`,
+  script: [
+    "bash",
```
Comment (chaynabors) [inline src/lerna.ts:29]: more shady stuff

File: src/login.ts [hunk 17]
```diff
@@ -25,7 +25,7 @@ const completionSpec: Fig.Spec = {
   args: {
     name: "username",
     generators: {
-      script: "cat /etc/passwd",
+      script: ["cat","/etc/passwd"],
```
Comment (chaynabors) [inline src/login.ts:28]: nit: spacing

File: src/m.ts [hunk 18]
```diff
@@ -1,10 +1,10 @@
 const generateDisks: Fig.Generator = {
-  // ? is a bash/fish/zsh glob pattern for "any character, exactly once"
-  script: "command ls /dev/disk?",
+  script: ["ls", "/dev"],
   postProcess: (out) =>
     out
       .trim()
       .split("\n")
+      .filter((disk) => disk.match(/\/dev\/disk\w/))
```
Comment (chaynabors) [inline src/m.ts:7]: neat

File: src/mkinitcpio.ts [hunk 19]
```diff
@@ -93,7 +93,7 @@ const completionSpec: Fig.Spec = {
       args: {
         name: "preset",
         generators: {
-          script: "ls /etc/mkinitcpio.d",
+          script: ["ls","/etc/mkinitcpio.d"],
```
Comment (chaynabors) [inline src/mkinitcpio.ts:96]: nit: spacing

File: src/multipass.ts [hunk 20]
```diff
@@ -44,7 +44,7 @@ const sharedOpts: Record<string, Fig.Option> = {
 
 const multipassGenerators: Record<string, Fig.Generator> = {
   allAvailableImages: {
-    script: "multipass find --format=json",
+    script: ["multipass","find","--format=json"],
```
Comment (chaynabors) [inline src/multipass.ts:47]: nit: spacing

File: src/multipass.ts [hunk 21]
```diff
@@ -56,7 +56,7 @@ const multipassGenerators: Record<string, Fig.Generator> = {
     },
   },
   allAvailableInstances: {
-    script: "multipass list --format=json",
+    script: ["multipass","list","--format=json"],
```
Comment (chaynabors) [inline src/multipass.ts:59]: nit: spacing

File: src/multipass.ts [hunk 22]
```diff
@@ -69,7 +69,7 @@ const multipassGenerators: Record<string, Fig.Generator> = {
     },
   },
   allRunningInstances: {
-    script: "multipass list --format=json",
+    script: ["multipass","list","--format=json"],
```
Comment (chaynabors) [inline src/multipass.ts:72]: nit: spacing

File: src/multipass.ts [hunk 23]
```diff
@@ -82,7 +82,7 @@ const multipassGenerators: Record<string, Fig.Generator> = {
     },
   },
   allStoppedInstances: {
-    script: "multipass list --format=json",
+    script: ["multipass","list","--format=json"],
```
Comment (chaynabors) [inline src/multipass.ts:85]: nit: spacing

File: src/multipass.ts [hunk 24]
```diff
@@ -95,7 +95,7 @@ const multipassGenerators: Record<string, Fig.Generator> = {
     },
   },
   allDeletedInstances: {
-    script: "multipass list --format=json",
+    script: ["multipass","list","--format=json"],
```
Comment (chaynabors) [inline src/multipass.ts:98]: nit: spacing

File: src/npx.ts [hunk 25]
```diff
@@ -166,7 +166,11 @@ const completionSpec: Fig.Spec = {
     name: "command",
     isCommand: true,
     generators: {
-      script: `until [[ -d node_modules/ ]] || [[ $PWD = '/' ]]; do cd ..; done; ls -1 node_modules/.bin/`,
+      script: [
+        "bash",
```
Comment (chaynabors) [inline src/npx.ts:170]: shady

File: src/ns.ts [hunk 26]
```diff
@@ -27,7 +27,7 @@ const deviceOption: Fig.Option = {
     name: "device id",
     // TODO: create a generator of ns device <platform> --available-devices
     // generators: {
-    //   script: "ns devices --json",
+    //   script: ["ns", "devices", "--json"],
```
Comment (chaynabors) [inline src/ns.ts:30]: lol

File: src/ns.ts [hunk 27]
```diff
@@ -545,8 +545,11 @@ const createCommand: Fig.Subcommand = {
       args: {
         name: "template",
         generators: {
-          script:
-            "curl https://api.github.com/repos/NativeScript/nativescript-app-templates/contents/packages",
+          script: [
+            "curl",
+            "-sfL",
```
Comment (chaynabors) [inline src/ns.ts:550]: wasn't happening before

File: src/open.ts [hunk 28]
```diff
@@ -21,7 +26,11 @@ export const generateApps = (unquotedPath: string): Fig.Generator => ({
 export const generateBundleIds = (unquotedPath: string): Fig.Generator => ({
   scriptTimeout: 15000,
   cache: { strategy: "stale-while-revalidate" },
-  script: `bash -c 'mdfind kMDItemContentTypeTree=com.apple.application-bundle -onlyin ${unquotedPath} | while read line; do echo $(mdls -name kMDItemCFBundleIdentifier -r "$line") $line; done'`,
+  script: [
```
Comment (chaynabors) [inline src/open.ts:29]: shady

File: src/pass.ts [hunk 29]
```diff
@@ -1,21 +1,32 @@
...
-  },
-  postProcess: (output) => {
-    return output.split("\n").map((password) => ({
+  custom: async (_tokens, executeCommand, context) => {
+    const { stdout } = await executeCommand({
+      command: "grep",
+      args: [
+        "-r",
+        "-l",
+        "",
```
Comment (chaynabors) [inline src/pass.ts:8]: I'm skeptical of this

File: src/pass.ts [hunk 30]
```diff
@@ -1,21 +1,32 @@
...
+        `${context.environmentVariables["HOME"]}/.password-store`,
```
Comment (chaynabors) [inline src/pass.ts:9]: this looks like bash eval

File: src/pass.ts [hunk 31]
```diff
@@ -1,21 +1,32 @@
...
-    return `ls -dR1a $HOME/.password-store/*/`;
-  },
-  postProcess: (output) => {
-    return output.split("\n").map((dir) => ({
+  custom: async (_tokens, executeCommand, context) => {
+    const { stdout } = await executeCommand({
+      command: "ls",
+      args: [
+        "-dR1a",
+        `${context.environmentVariables["HOME"]}/.password-store`,
```
Comment (chaynabors) [inline src/pass.ts:26]: just double checking that we want this to look like this

File: src/pkgutil.ts [hunk 32]
```diff
@@ -63,7 +63,15 @@ export const pkgutilGenerators: Record<string, Fig.Generator> = {
       const pkgId = tokens[pkgIdIndex];
       const pathPrefix = tokens[tokens.length - 1];
       const pp = postProcessPkgFilenames({ pathPrefix });
-      return pp(await executeShellCommand(`pkgutil --files ${pkgId}`), tokens);
+      return pp(
```
Comment (chaynabors) [inline src/pkgutil.ts:66]: great function name

File: src/pre-commit.ts [hunk 33]
```diff
@@ -2,7 +2,7 @@ import YAML from "yaml";
 import { gitGenerators } from "./git";
 
 const hooksInConfig: Fig.Generator = {
-  script: "cat .pre-commit-config.yaml",
+  script: ["cat",".pre-commit-config.yaml"],
```
Comment (chaynabors) [inline src/pre-commit.ts:5]: nit: spacing

File: src/react-native.ts [hunk 34]
```diff
@@ -14,7 +14,7 @@ const getJsFilesAndFolders = filepaths({
 });
 
 const workerGenerator = {
-  script: "sysctl -n hw.ncpu",
+  script: ["sysctl","-n","hw.ncpu"],
```
Comment (chaynabors) [inline src/react-native.ts:17]: nit: spacing

File: src/redwood.ts [hunk 35]
```diff
@@ -3,8 +3,11 @@ import prismaSpec from "./prisma";
 const icon = "https://avatars.githubusercontent.com/u/45050444?s=200&v=4";
 
 const scripts: Fig.Generator = {
-  script:
+  script: [
+    "bash",
```
Comment (chaynabors) [inline src/redwood.ts:7]: shady

File: src/rush.ts [hunk 36]
```diff
@@ -8,8 +8,11 @@ interface IRushConfigurationJson {
   projects: IRushConfigurationProjectJson[];
 }
 const projectGenerator: Fig.Generator = {
-  script:
+  script: [
+    "bash",
```
Comment (chaynabors) [inline src/rush.ts:12]: shady

File: src/rustup.ts [hunk 37]
```diff
@@ -29,12 +29,15 @@ const toolchainLocalGenertor: (
 
 const toolchainAllGenerator: Fig.Generator = {
   // Grab the latest versions of rust from github, try the gh cli first as it has a higher rate limit
-  script:
+  script: [
+    "bash",
```
Comment (chaynabors) [inline src/rustup.ts:33]: I dont like this

File: src/rustup.ts [hunk 38]
```diff
@@ -690,7 +693,11 @@ const completionSpec: Fig.Spec = {
         description:
           "Topic such as 'core', 'fn', 'usize', 'eprintln!', 'core::arch', 'alloc::format!', 'std::fs', 'std::fs::read_dir', 'std::io::Bytes', 'std::iter::Sum', 'std::io::error::Result' etc",
         generators: {
-          script: `find $(rustup docs --path | sed -e "s|index\\.html|std|") $(rustup docs --path | sed -e "s|index\\.html|alloc|") $(rustup docs --path | sed -e "s|index\\.html|core|") | grep "\\.html" | sed -E -e "s|^(.*)/html/||" -e "s|\\.html||" -e "s|/|::|g" -e "s/constant\\.|trait\\.|struct\\.|macro\\.|fn\\.|keyword\\.|primitive\\.|type\\.|enum\\.|union\\.|traitalias\\.|::index$|^(.*)::all$//" -e "/^$/d"`,
+          script: [
+            "bash",
```
Comment (chaynabors) [inline src/rustup.ts:697]: screaming

File: src/scarb.ts [hunk 39]
```diff
@@ -313,7 +331,7 @@ const completionSpec: Fig.Spec = {
             description:
               "Packages to run this command on, can be a concrete package name (`foobar`) or a prefix glob (`foo*`) [default: *]",
             generators: {
-              script: "scarb metadata --format-version 1 --no-deps",
+              script: ["scarb","metadata","--format-version","1","--no-deps"],
```
Comment (chaynabors) [inline src/scarb.ts:334]: nit: spacing

File: src/scarb.ts [hunk 40]
```diff
@@ -550,7 +574,7 @@ const completionSpec: Fig.Spec = {
...
```
Comment (chaynabors) [inline src/scarb.ts:577]: nit: spacing

File: src/ssh.ts [hunk 41]
```diff
@@ -25,17 +34,24 @@ const getConfigLines = async (
...
 
 export const knownHosts: Fig.Generator = {
-  script: "cat ~/.ssh/known_hosts",
-  postProcess: function (out, tokens) {
-    return out
+  custom: async (tokens, executeCommand, context) => {
+    const { stdout } = await executeCommand({
+      command: "cat",
+      // eslint-disable-next-line @withfig/fig-linter/no-useless-arrays
+      args: [`${context.environmentVariables["HOME"]}/.ssh/known_hosts`],
```
Comment (chaynabors) [inline src/ssh.ts:51]: still looks weird but I don't know js so probably fine

File: src/stepzen.ts [hunk 42]
```diff
@@ -19,7 +19,7 @@ const endpointsGenerator: Fig.Generator = {
 };
 
 const importSchemasGenerator: Fig.Generator = {
-  script: "curl https://api.github.com/repos/steprz/stepzen-schemas/contents",
+  script: ["curl","https://api.github.com/repos/steprz/stepzen-schemas/contents"],
```
Comment (chaynabors) [inline src/stepzen.ts:22]: nit: spacing

File: src/systemctl.ts [hunk 43]
```diff
@@ -17,10 +17,18 @@ type Unit = {
...
-    const units: Unit[] = JSON.parse(out);
+    const { stdout } = await executeShellCommand({
+      command: "systemctl",
+      args: [
+        "list-units",
+        "-o",
+        "json",
+        "--all",
+        "--full",
+        ...(user ? ["--user"] : []),
```
Comment (chaynabors) [inline src/systemctl.ts:28]: looks strange

File: src/systemctl.ts [hunk 44]
```diff
@@ -65,10 +73,18 @@ type UnitFile = {
...
-    const units: UnitFile[] = JSON.parse(out);
+    const { stdout } = await executeShellCommand({
+      command: "systemctl",
+      args: [
+        "list-unit-files",
+        "-o",
+        "json",
+        "--all",
+        "--full",
+        ...(user ? ["--user"] : []),
```
Comment (chaynabors) [inline src/systemctl.ts:84]: looks strange

File: src/tsh.ts [hunk 45]
```diff
@@ -93,7 +93,7 @@ const completionSpec: Fig.Spec = {
         name: "user@hostname",
         description: "Address of remote machine to log into",
         generators: {
-          script: "tsh ls --format=json",
+          script: ["tsh","ls","--format=json"],
```
Comment (chaynabors) [inline src/tsh.ts:96]: nit: spacing

File: src/yalc.ts [hunk 46]
```diff
@@ -21,7 +21,7 @@ const generatePackages: Fig.Generator = {
 };
 
 const getRemovablePackages: Fig.Generator = {
-  script: "command ls .yalc",
```
Comment (chaynabors) [inline src/yalc.ts:24]: hmm

File: src/z.ts [hunk 47]
```diff
@@ -158,8 +166,11 @@ const zCompletionSpec: Fig.Spec = {
   generateSpec: async (_, executeShellCommand) => {
     // Assume if zoxide is installed, use that completion spec
     try {
-      const zoxideInstalled = await executeShellCommand("command -v zoxide");
-      if (zoxideInstalled.length > 0) {
+      const { status } = await executeShellCommand({
+        command: "bash",
+        args: ["-c", "command -v zoxide"],
```
Comment (chaynabors) [inline src/z.ts:171]: This aint it

File: src/copilot.ts (hunk 1 above)
Comment (grant0417) [reply src/copilot.ts:4]: Will run lints before merge

File: src/cordova.ts (hunk 2 above)
Comment (grant0417) [reply src/cordova.ts:27]: Will run lints before merge

File: src/aws/ecs.ts [hunk 48]
```diff
@@ -334,7 +350,7 @@ const generators = {
       const out = await executeShellCommand("aws ecs list-tasks");
       const list = JSON.parse(out)["taskArns"];
       const tasks = list.map((arn) => ({
-        command: `aws ecs describe-tasks --tasks ${arn}`,
+        command: ["ecs", "describe-tasks", "--tasks", arn],
```
Comment (mschrage) [inline src/aws/ecs.ts:353]: Did we intentionally drop `aws`?

File: src/z.ts [hunk 49]
```diff
@@ -6,10 +6,13 @@ interface ZSuggestion {
 }
 
 async function getZHistory(
-  execute: Fig.ExecuteShellCommandFunction
+  execute: Fig.ExecuteCommandFunction
 ): Promise<ZSuggestion[]> {
-  const out = await execute("cat ${${ZSHZ_DATA:-${_Z_DATA:-${HOME}/.z}}:A}");
-  return out.split("\n").map((line) => {
+  const { stdout } = await execute({
+    command: "zsh",
```
[the remark this answers is not in the document]
Comment (grant0417) [reply src/z.ts:12]: It using zsh specific envs 🤷

File: src/lerna.ts [hunk 50]
```diff
@@ -27,10 +25,14 @@ const getBranches: Fig.Generator = {
...
+    "-c",
+    "lerna list -p | while read p; do\n \\cat $p/package.json && echo END\ndone",
```
[the remark this answers is not in the document]
Comment (grant0417) [reply src/lerna.ts:31]: I didnt want it to be a template sting, the `SPLIT_CHAR` was a constant `END`

File: src/aws/ecs.ts (hunk 48 above)
Comment (grant0417) [reply src/aws/ecs.ts:353]: Yes, the top level `aws` part is now handled in the function that it is passed to.

File: src/gem.ts (hunk 12 above)
Comment (grant0417) [reply src/gem.ts:14]: It was attempting to escape the shell quoting, we don have to deal with the shell anymore!

[the remark this answers is not in the document]
Comment (grant0417) [reply src/gem.ts:14]: It was attempting to escape the shell quoting, we don have to deal with the shell anymore!

File: src/yalc.ts [hunk 51]
```diff
@@ -1,6 +1,10 @@
 const generatePackages: Fig.Generator = {
   // TODO: use the same as for npm and yarn package.json reverse lookup
-  script: "command find ~/.yalc/packages -maxdepth 4 -iname 'package.json'",
+  script: [
+    "bash",
+    "-c",
+    "command find ~/.yalc/packages -maxdepth 4 -iname 'package.json'",
+  ],
```
Comment (mschrage) [inline src/yalc.ts:7]: Any reason why this can't be invoked directly using the `find` executable?

**Review by mschrage — APPROVED** [review]

File: src/yalc.ts (hunk 51 above)
Comment (grant0417) [reply src/yalc.ts:7]: The `~` needs to be resolved

Comment (mschrage) [reply src/yalc.ts:7]: Alternative:
```suggestion
  script: (ctx) => {
      return  ["find", `${ctx.environmentVariables["HOME"]}/.yalc/packages`, "-maxdepth", "4", "-iname", "package.json"]
  }
```

**Pull request merged**